scikeras/__init__.py

```diff
@@ -3,12 +3,18 @@
 __author__ = """Adrian Garcia Badaracco"""
 __version__ = "0.2.0"
 
+from tensorflow import keras
+from tensorflow.python.keras import metrics as _metrics  # noqa
 
-# Monkey patch log_cosh reference
-# See https://github.com/tensorflow/tensorflow/pull/42097
-# Will be removed whenever the
-# min supported version of tf incorporates the fix
-from tensorflow.python import keras  # noqa
 
+_metrics.log_cosh = (
+    _metrics.logcosh
+)  # See https://github.com/tensorflow/tensorflow/pull/42097
 
-keras.metrics.log_cosh = keras.metrics.logcosh
+from scikeras import _saving_utils as saving_utils  # noqa
+
+
+keras.Model.__reduce__ = saving_utils.pack_keras_model
+keras.losses.Loss.__reduce__ = saving_utils.pack_keras_loss
+keras.metrics.Metric.__reduce__ = saving_utils.pack_keras_metric
+keras.optimizers.Optimizer.__reduce__ = saving_utils.pack_keras_optimizer
```

Comment:
Shouldn't these lines go in the definitions of Model/Loss/etc?

Reply:
Do you mean in the class def within Tensorflow? Maybe I'm not understanding...

Reply:
Oh, I see. These are *Keras* classes, not SciKeras classes. Then why not put this into a Keras RFC?

Reply:
Right now, these things are very hacky. The optimizer hack is never going to make it past any sort of review (I'm not even sure it works for 100% of cases; it relies on private methods). `Model.__reduce__` does a bunch of zip file stuff that is also pretty hacky, and really should be replaced with TF utilities that need to be implemented on the C++ backend (I think) as discussed in the TF PR (https://github.com/tensorflow/tensorflow/pull/39609#issuecomment-690544525).

Reply:
I think there's two levels for RFCs: the goal and how to achieve that goal.

This PR shows the goal is feasible and practical, and serves as a prototype (`_restore_optimizer_weights` is clearly a rough prototype). It's clear what needs to be implemented in a cleaner way. Inside TF, there's no downside to depending on private functions.

Reply:
I agree that this PR shows that it's feasible, but a (similar) thing is already at https://github.com/tensorflow/tensorflow/pull/39609.

I guess, let me just ask, what steps do you suggest be taken? Open a new RFC or PR the optimizer weights part specifically?

Reply:
I might open a new PR to TF after this PR is merged. Or maybe add it to that PR with an explaining comment if the implementation is simple.

If they want an RFC they'll call for it.

Reply:
I have no problem opening a hacky PR in TF just for visibility. Just to be clear, we're talking specifically about the optimizer weights part right (not about Metrics or Model)? And specifically, about the hack to restore the weights for `Adam` and other optimizers that use slots, _not_ about implementing `__reduce__` for all optimizers?

Reply:
> we're talking specifically about the optimizer weights part right

My concern is Keras having a working serialization method. I think Models and optimizers are most important use cases (plus all metrics I can think of are stateless past some rolling average).

I'm not sure what the difference is between "restoring weights for Adam" and "implementing `__reduce__`." Do they both accomplish a working serialization method? Do they have a different implementation?

Reply:
Yes, they are fundamentally two distinct things.

In this PR, we are fixing the weight restoration bug by implementing some hacks within `Model.__reduce__`. I reported the bug and showed how this hack works in https://github.com/tensorflow/tensorflow/issues/44670 in hopes that that will make it easier for them to fix the bug. To actually fix the bug, one would have to implement an actual fix within the TF `SaveModel` ecosystem. I tried and failed to do that. Thus I don't see that there's any further PRs I can submit to TensorFlow.

As for implementing `Model.__reduce__`, I think we should update the existing PR with whatever implementation we land on here, sans the hack to restore optimizer weights since that is functionally unrelated.

I hope this makes sense! I know it's a convoluted topic.